### ISSUE TITLE
option to filter out tiny/small/invisible features to speed-up rendering

### ISSUE THREAD

**Issue by stelf** [qgis/QGIS#33013]
**Feature description.**

QGIS should provide option to have the minimum size of feature being drawn (particularly when it comes to shapes) designated somehow by the user so that no irrelevant or invisible data is rendered. 

Note: the selection tool may still query all features (again - option).

Existence of such option may result in dramatic speedup of results loaded & rendered when navigating zoomed-out version of some large map (think 100k+ distinct features, typically the case with city-level maps).

What is going to be achieved is several things :

* less data moved from DB to QGIS
* less effort rendering data
* clearer representation of some maps, that feature large number of small, nearly-invisible, or overlapping features when approached 'from above' 

Considerations:

* calculate the effective size of bounding box that a feature should go out in respect to minimum number of pixels it is mapped into on the screen (1, 4, 9) and never load such features
* designate the minimum size (in shape's coordinate system) that should be drawn
* designate whether search takes into consideration invisible shapes or not

**Additional context**

It is absolutely useless to draw data that is not going to be comprehended. Or data that overlaps in respect to certain screen resolution. Even if there's a way to feature out such data with existing QGIS tools, rows and geometry data is still transferred between RDBMS and QGIS.

**Comment by Saijin-Naib** [comment]
This seems great, and perhaps could be tied into a progressive-quality rendering system like in Issue #30315

**Comment by stelf** [comment]
now that I think about it - maybe is good for grant, as it possibly benefits (in terms of speed, effective work) everyone ever touching QGIS.

**Comment by roya0045** [comment]
To be honest we have something that could possibly be altered. I haven't looked at the code yet but if we consider the labeling parameter that lets your not draw a label for feature smaller than some size, I guess something similar could be done. 

But I'm not sure about how it works internally without checking, it could either render and then evaluate the size or evaluate the size and then render by converting the size to mm in the case of mm based threshold.

If its the second case it should be reusable for rendering to some extent. Though once again I'm not sure if we can fiddle the the rendering mechanic or if everything is handled internally in qt.

**Comment by stelf** [comment]
Rendering is one point, moving millions of features from the RDBMS is another, and even with extent applied on the SQL level, it loads features that are barely if ever visible at certain zoom. with more than 100k features - this is absolutely perceivable and keeps loads on the DB. In the case with PostGIS not that big an issue, provided that everything is properly indexed, not sure with Spatialite, and cases when indices are missing.

**Comment by roya0045** [comment]
> Rendering is one thing, moving millions of features from the RDBMS another. Even with extent applied on the SQL level, it loads features that are barely if ever visible at certain zoom. with more than 100k features - this is absolutely perceivable and keeps loads on the DB. In the case with PostGIS not that big an issue, provided that everything is properly indexed, not sure with Spatialite, and cases when indices are missing.

Exactly, I don't think there is a magic way to implement this, either make the db calculate the size/area and keep it as a column to use in the query or filter them before rendering when you can't make a query to select only the big ones within the extent.

**Comment by stelf** [comment]
I have been thinking more about this recently, and there's another perspective that should be considered, maybe, generally, by the GIS community. 

People seldom need the data to be correct when they see it. Actually - no one really cares, as long as the measurement tools and snapping works. And even then - one relies not on the visual aid of correct and complete data being shown, but rather on the usability of the software and its ability to correctly measure distances. 

Thus, the understanding that it is imperative that a map be rendered correctly is inherently wrong, as the map render is only visual aid, a view to the data, and it is important that this aid is indeed aid and not a burden, which is particularly true with large datasets where hundreds of thousands of features potentially show/overlap/etc. 

A mere human has no way to neither tell whether this data is complete, nor if it's correct. So it is unreasonable to try to draw all that there is in the storage. This understanding may be correct for paper maps, printed out stuff, but the magic of screens and computers is not about bringing more data - should be about showing as much data as needed to visually aid the person navigating in this data without causing burnt CPUs, network delays, etc. 

QGIS being targeted at academia may easily pioneer this approach. Although it is to my surprise academia that overlooks the human side of interfaces and goes well along with attribute tables showing a NxM matrix of inhumane volume of information, just for the sake of ... showing this data. Or trying to correctly render inhumanely large amounts of data with number of features easily exceeding the number of pixels on a screen. 

Computers should aid us in navigating the data. Solving the challenge of showing that much data on this small screen is for academic papers, not for users. 

@Saijin-Naib
